File: ruby/sieve/sieve.rb
# Sieve of Eratosthenes to find prime numbers
class Sieve
  attr_reader :primes

  def initialize(limit)
    @limit = limit
    @primes = []
    eratosthenes if limit > 1
  end

  private

  def eratosthenes
    @primes = (2..@limit).to_a
    remove_multiples(2)
    Range.new(3, Math.sqrt(@limit).floor).step(2) do |i|
      remove_multiples(i) unless Math.sqrt(i).integer?
    end
  end

  def remove_multiples(num)
    step = num * (num == 2 ? 1 : 2)
    Range.new(num**2, @limit).step(step) { |i| @primes.delete(i) }
  end
end

File: python/sieve/sieve.py
def primes(limit):
    if limit < 2:
        return []

    # initialize the candidates array
    candidates = list(range(limit+1))
    candidates[0] = None
    candidates[1] = None

    # first prime
    remove_multiples_of(2, candidates, limit)

    # subsequent primes
    for i in range(3, limit, 2):
        if candidates[i] is not None:
            remove_multiples_of(i, candidates, limit)

    # return the prime numbers
    return [n for n in candidates if n is not None]


def remove_multiples_of(n, candidates, limit):
    step = n * (1 if n == 2 else 2)
    for i in range(n*n, limit+1, step):
        candidates[i] = None
